File: brainfuck2/bf.py
import platform
import sys, time
import os

INC = 1
MOVE = 2
LOOP = 3
PRINT = 4

class Op(object):
    def __init__(self, op, val):
        self.op = op
        self.val = val

class Tape(object):
    def __init__(self):
        self.tape = [0]
        self.pos = 0
    def get(self):
        return self.tape[self.pos]
    def inc(self, x):
        self.tape[self.pos] += x
    def move(self, x):
        self.pos += x
        while self.pos >= len(self.tape):
            self.tape.append(0)

def parse(iterator):
    res = []
    while True:
        try:
            # this check does not affect on performance
            if sys.version >= "3":
                c = iterator.__next__()
            else:
                c = iterator.next()
        except StopIteration:
            break

        if c == "+":
            res.append(Op(INC, 1))
        elif c == "-":
            res.append(Op(INC, -1))
        elif c == ">":
            res.append(Op(MOVE, 1))
        elif c == "<":
            res.append(Op(MOVE, -1))
        elif c == ".":
            res.append(Op(PRINT, 0))
        elif c == "[":
            res.append(Op(LOOP, parse(iterator)))
        elif c == "]":
            break

    return res

def _run(program, tape):
    for op in program:
        if op.op == INC: tape.inc(op.val)
        elif op.op == MOVE: tape.move(op.val)
        elif op.op == LOOP: 
            while tape.get() != 0:
                _run(op.val, tape)
        elif op.op == PRINT:
            sys.stdout.write(chr(tape.get()))
            sys.stdout.flush()

class Program(object):
    def __init__(self, code):
        self.ops = parse(iter(code))

    def run(self):
        _run(self.ops, Tape())

import os
t = time.time()
f = open(sys.argv[1], "r")
text = f.read()

Program(text).run()
sys.stderr.write("time({0})\n".format(time.time() - t))

File: brainfuck2/bf.rb

class Op
  attr_accessor :op, :val
  def initialize(op, val)
    @op = op
    @val = val
  end
end

class Tape
  def initialize
    @tape = [0]
    @pos = 0
  end

  def get
    @tape[@pos]
  end

  def inc(x)
    @tape[@pos] += x
  end

  def move(x)
    @pos += x
    while (@pos >= @tape.size)
      @tape << 0
    end
  end
end

class Program
  def initialize(code)
    @ops = parse code.chars.each
  end

  def run
    _run @ops, Tape.new
  end

private

  def _run(program, tape)
    program.each do |op|
      case op.op
        when 0; tape.inc(op.val)
        when 1; tape.move(op.val)
        when 3; _run(op.val, tape) while tape.get != 0
        when 2; print(tape.get.chr)
      end
    end
  end

  def parse(iterator)
    res = []
    while c = iterator.next rescue nil
      op = case c
           when '+'; Op.new(0, 1)
           when '-'; Op.new(0, -1)
           when '>'; Op.new(1, 1)
           when '<'; Op.new(1, -1)
           when '.'; Op.new(2, 0)
           when '['; Op.new(3, parse(iterator))
           when ']'; break
           end
      res << op if op
    end
    res
  end
end

def run(text)
  t = Time.now
  Program.new(text).run
  STDERR.puts("time(#{Time.now - t})")
end

text = IO.read(ARGV[0])
times = (ARGV[1] || 1).to_i
STDERR.puts("started")
times.times { run(text) }
